### PR TITLE
fix: make "find references" multi-token mapping aware

### PR DESCRIPTION
**Pull request by jonas-schievink**
Part of https://github.com/rust-analyzer/rust-analyzer/issues/10070

I was hoping that this would fix "find references" on salsa queries, but salsa emits multiple defs sharing the same span (the trait method, and an impl of that trait).

### CODE REVIEW

File: crates/ide/src/references.rs [hunk 1]
```diff
@@ -99,29 +100,40 @@ pub(crate) fn find_def(
...
-            NameClass::PatFieldShorthand { local_def, field_ref: _ } => {
-                Definition::Local(local_def)
-            }
-        },
-        ast::NameLike::Lifetime(lifetime) => NameRefClass::classify_lifetime(sema, &lifetime)
-            .and_then(|class| match class {
-                NameRefClass::Definition(it) => Some(it),
-                _ => None,
+    // FIXME: our macro handling here could be improved, ideally we'd disambiguate if the token maps
+    // to multiple defs, but LSP doesn't really allow that.
```
Comment (matklad) [inline crates/ide/src/references.rs:104]: We don’t care about LSP at this layer, we can (and should) return whatever makes most sense, and dumb it down in to_proto if necessary.

**Comment by matklad** [comment]
Can we add a test(ing infra)? As our proc macros are behind a trait, I think what we can do is write a fake impl of that trait, which doesn’t have to deal with dylibs and stuff.


Actually, do we already have such fake expanders? It feels very obvious that we should do this already for all kinds of tests, and, if we are not doing that, that seems like a broken window to me

**Comment by jonas-schievink** [comment]
Right, all the stuff dealing with multi-token mapping is also relevant for MBE, we don't need proc macros there. I do still plan on adding some general proc macro tests though.

File: crates/ide/src/references.rs (hunk 1 above)
Comment (Veykril) [reply crates/ide/src/references.rs:104]: Ye we should adjust this to handle multiple defs. I also think we should show the references for all definitions a token refers to if a macro uses the span for multiple defs when converting to proto.
